src/reconcile/rec_visibility.c

```diff
@@ -152,16 +152,16 @@ __rec_append_orig_value(
  *     Return if we need to save the update chain
  */
 static bool
-__rec_need_save_upd(WT_SESSION_IMPL *session, WT_UPDATE *selected_upd, uint64_t max_txn,
-  wt_timestamp_t max_ts, bool has_newer_updates, uint64_t flags, WT_PAGE *page)
+__rec_need_save_upd(WT_SESSION_IMPL *session, WT_UPDATE_SELECT *upd_select, bool has_newer_updates,
+  uint64_t flags, WT_PAGE *page)
 {
     /*
      * Save updates for any reconciliation that doesn't involve history store (in-memory database
      * and fixed length column store), except when the maximum timestamp and txnid are globally
      * visible.
      */
     if (LF_ISSET(WT_REC_IN_MEMORY) || page->type == WT_PAGE_COL_FIX)
-        return (!__wt_txn_visible_all(session, max_txn, max_ts));
+        return (!__wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts));
```

Comment:
Why we only need to check the visibility for on page update here? I don't think this line of change is correct. Here we want to make sure every update on the update chain is visible to all so that we can safely discard the update chain.

May be change it to
```
if (LF_ISSET(WT_REC_IN_MEMORY) || page->type == WT_PAGE_COL_FIX) {
        if (has_newer_updates)
               return true;
        return (!__wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts));
}
```
before this section?

Reply:
Thanks @quchenhao, I think you're right.  I found this function hard to follow, so I've tried to simplify it in ba67e79: can you please take a look and let me know if this addresses your concerns?

Happy to discuss if it isn't clear or I've misunderstood.